### DISCUSSION TITLE
Siemens PLC gives BadTooManyOperations

### DISCUSSION THREAD

**Discussion by Randelung** [FreeOpcUa/opcua-asyncio#543] [Q&A] [answered]
Hey guys

I wanted to ask if you have stumbled upon this error before. I'm trying to connect to a Siemens PLC S7-1517 with a little more than 1100 nodes in (currently) one subscription. According to [the FAQ](https://support.industry.siemens.com/cs/document/109755846/what-are-the-system-limits-of-the-opc-ua-server-with-s7-1500-and-s7-1200-?dti=0&lc=en-US) it should be performant enough to handle that.
However, I'm getting BadTooManyOperations, which asyncua logs as a service fault in response to the CreateMonitoredItemsRequest.
```
2021-04-08 14:12:00.101 WARNING  asyncua.client.ua_client.UASocketProtocol: ServiceFault from server received  in response to CreateMonitoredItemsRequest
```
![image](https://user-images.githubusercontent.com/3729270/114023932-c04f1000-9873-11eb-86b5-261e50128ef5.png)
![image](https://user-images.githubusercontent.com/3729270/114024009-d230b300-9873-11eb-952b-e2023c4e451e.png)

I'm thinking the issue may be the limited capabilities of the PLC in combination with the combined CreateMonitoredItemsRequest and PublishRequest. However, it did work when subscribing fewer than 1000 nodes, so it might be something else. From the FAQ I'm assuming that it handles subscriptions with more than 1000 nodes differently, so that may be the reason why it worked with fewer nodes... idk.

Is it possible to suppress or delay this first PublishRequest?
Any other insight?

Thanks for your input.

**Comment by AndreasHeine** [comment, marked as answer]
https://reference.opcfoundation.org/v104/Core/docs/Part4/5.12.2/

>    itemsToCreate []

how many items in the call?
have you tryed to subscribe just a chunk and in another call to add the rest?

**Reply by AndreasHeine** [reply]
should be in CreateMonitoredItemsRequest!

**Reply by AndreasHeine** [reply]
![Unbenannt](https://user-images.githubusercontent.com/56362817/114026896-15d8ec00-9877-11eb-8fa8-6f7c0e81c1a5.PNG)

**Reply by Randelung** [reply]
Damn, you're fast, I appreciate it.

![image](https://user-images.githubusercontent.com/3729270/114026964-2f7a3380-9877-11eb-857f-25daa4a69135.png)
1131 items.

I didn't try to split the load, no. I'm looking at the debug output of the PLC with a collegue, too, which is telling me that I've apparently run through some limitation.
![image](https://user-images.githubusercontent.com/3729270/114027253-7700bf80-9877-11eb-9524-b543dc89eb44.png)
The message isn't very helpful. We're not finding any settings that would be problematic, either, max monitored items was set to 10k.

I don't think the number of nodes is the issue since it's not a BadTooManyMonitoredItems error as specified in the FAQ.

**Reply by Randelung** [reply]
Oh that's good, so there's a setting somehwere... or I'm misinterpreting the FAQ.

**Reply by AndreasHeine** [reply]
check with ua expert
![Unbenannt1](https://user-images.githubusercontent.com/56362817/114028589-efb44b80-9878-11eb-8e66-62bd39a26a41.PNG)

tia blames propably the limit per call

**Reply by Randelung** [reply]
Yeah, I did confirm the 1000 you showed. But we can't find a setting to change it, sadly.
I guess I'll have to go with several subscriptions.
Thanks for your help!

**Reply by Randelung** [reply]
I'm kind of irritated that the error code explicitly mentioned on the FAQ is not the one sent in this exact case.

**Reply by AndreasHeine** [reply]
just read the MaxMonitoredItemsPerCall if you list of nodes is bigger split the list in half and create with the first half and subscribe the second half in a different request after the first half is subscribed!

**Reply by AndreasHeine** [reply]
```python
    client = Client(url='opc.tcp://192.168.10.1:4840/')
    async with client:
        handler = SubscriptionHandler()
        subscription = await client.create_subscription(2000, handler)

        nodes1 = await client.get_node(ua.ObjectIds.Server_ServerStatus).get_children()
        nodes2 = await client.get_node(ua.ObjectIds.Server_ServerStatus).get_children()
        nodes3 = await client.get_node("ns=3;s=PLC").get_children()
        nodes4 = await client.get_node("ns=3;s=DataBlocksGlobal").get_children()
        nodes5 = await client.get_node("ns=3;s=DataBlocksGlobal").get_children()
        nodes6 = await client.get_node("ns=3;s=Inputs").get_children()
        nodes7 = await client.get_node("ns=3;s=Memory").get_children()
        nodes8 = await client.get_node("ns=3;s=Outputs").get_children()

        # 1200+ nodes
        nodes = nodes1 + nodes2 + nodes3 + nodes4 + nodes5 + nodes6 + nodes7 + nodes8
        print("Len nodes: ", len(nodes))

        maxpercall = await client.get_node("i=11714").read_value() # 1000 nodes
        chunks = int(len(nodes)/maxpercall)
        chunks += 1
        print(chunks)
        nodes_to_sub = []
        if chunks == 1:
            nodes_to_sub = nodes
            await subscription.subscribe_data_change(nodes_to_sub)
        if chunks > 1:
            i = 0
            while 1:
                i+=1
                nodes_to_sub = nodes[:maxpercall]
                print(i, "Subscribing to:", len(nodes_to_sub), "Nodes")
                await subscription.subscribe_data_change(nodes_to_sub)
                del nodes[:maxpercall]
                if len(nodes) <= 0:
                    break
```

**Reply by Randelung** [reply]
I did
```python
        for chunk in [nodes[i:i + chunk_size] for i in range(0, len(nodes), chunk_size)]:
            results = await self.__subscription.subscribe_data_change(chunk)
            for i in range(len(results)):
                if isinstance(results[i], asyncua.ua.StatusCode):
                    self.__logger.error(f'{chunk[i].nodeid} not subscribed')
                    raise ConnectionError(f'{chunk[i].nodeid} not subscribed')
```
which seems to work.

**Reply by Randelung** [reply]
I think I understood my brain fart (and I'm writing this for posterity, because I'm pretty sure you know already 😄).
The call limit is not the subscription limit outlined in the FAQ. Obviously I was able to incrementally add more monitored items to the subscription and "a subscription with a maximum of 1000 elements cannot be converted into a subscription with more than 1000 elements" apparently wasn't triggered.
I thought that in order to avoid this limitation I had to subscribe to all of the nodes at once, which turned out to be untrue.
I'm also guessing that the BadTooManyMonitoredItems fault would be triggered only if the 10'000 items, or whatever is configured on the PLC, is surpassed.

Thanks again for your help.

**Reply by starturtle** [reply]
It says max items per _call_; mayhaps something in the server can't handle such a large request payload. Splitting sounds like a great idea. Max monitored items is the amount the server can hold, not the amount you can request in one go, afaicr.

**Reply by Randelung** [reply]
Yeah, my mistake was thinking in order to avoid the <1000 / >1000 conversion mentioned in the FAQ I'd have to subscribe all at once.

**Reply by AndreasHeine** [reply]
its best practice that the client reads some server data up front (capabilities, status, ...) and adopt to the server by only asking things which the server is able to handle otherwise you will always try catch that service faults (waist of server-ressources just for failing requests in my oppinion). a important part in opc ua is negotiation between server and client (like maxmessagsize, maxchuncksize, revised intervals, ...) but also how many nodes can i browse with one request at once, ... !
